### ISSUE TITLE
`self.all_gather` in test_epoch_end hangs forever

### ISSUE THREAD

**Issue by ecolss** [Lightning-AI/lightning#16541]
### Bug description

See the python code below, `all_gather` and `is_global_zero` make the program hang forever, no other error messages.
If `is_global_zero` is removed, it can finish successfully.

- pytorch_lightning version: 1.8.3.post1
- torch version: 1.10.1+cu113
- linux OS

### How to reproduce the bug

```python
import pytorch_lightning as thl
import torch as th
from torch.utils.data import Dataset, DataLoader, TensorDataset
import torch.nn.functional as F


class DummyModel(thl.LightningModule):
    def __init__(self, dim_in, dim_out):
        super().__init__()
        self.fc = th.nn.Linear(dim_in, dim_out)

    def forward(self, x):
        return self.fc(x)

    def configure_optimizers(self):
        return th.optim.Adam(self.parameters())

    def training_step(self, batch, idx):
        x, y = batch
        logits = self.forward(x)
        return F.binary_cross_entropy_with_logits(logits, y)

    def test_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.fc(x)
        return  y_hat

    def test_epoch_end(self, outputs):
        if self.trainer.is_global_zero:
            outputs = self.all_gather(outputs)
            print(f"after gather, len = {len(outputs)}")
            return outputs

if __name__ == "__main__":
    x = th.rand(10, 5)
    y = th.rand(10, 1)
    dataset = TensorDataset(x, y)
    loader = DataLoader(dataset, batch_size=5)

    model = DummyModel(5, 1)
    trainer = thl.Trainer(max_epochs=1, accelerator="gpu", devices=[2,3], strategy="ddp")
    trainer.fit(model, loader)
    res = trainer.test(model, dataloaders=loader)
    print(res)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @borda @awaelchli

**Comment by akihironitta** [comment]
I believe `all_gather` needs to be called not only in the 0th process but also all processes:

```diff
     def test_epoch_end(self, outputs):
+        outputs = self.all_gather(outputs)
         if self.trainer.is_global_zero:
-            outputs = self.all_gather(outputs)
             print(f"after gather, len = {len(outputs)}")
-            return outputs
+        return outputs
```

**Comment by ecolss** [comment]
> I believe `all_gather` needs to be called not only in the 0th process but also all processes:

@akihironitta Thanks very much for your quick rely.
Hmm, interesting since I remember I saw somewhere an example code, which tells people how to use `is_global_zero ` and `all_gather` together, and I just followed that code.
If `all_gather` needs to be called on all processes, would it be better to add a note in the lightning API documentations?

**Comment by akihironitta** [comment]
@itsnamgyu Haven't looked into the detail yet, but what you could try as a quick workaround is to disable inference mode:

```python
# default used by the Trainer
trainer = Trainer(inference_mode=True)

# Use `torch.no_grad` instead
trainer = Trainer(inference_mode=False)
```

https://pytorch-lightning.readthedocs.io/en/1.9.3/common/trainer.html#inference-mode

**Comment by itsnamgyu** [comment]
Thanks for the super quick reply. I'll check it out.

**Comment by ecolss** [comment]
> @ecolss If you can find that example with the wrong information, we can fix it. This applies to all collectives. They cannot work unless all processes enter them and exchange data. Just think about what it actually means if only process 0 would enter the if statement: Where would the other data to be gathered come from, if the other processes don't know they should send it?
> 
> A documentation update that says: "This method needs to be called on all processes" or similar would be welcome :) (if we don't have it already)

@awaelchli I think your teammate delivered the wrong message here https://github.com/Lightning-AI/lightning/discussions/9259#discussioncomment-1858041

**Comment by bocopilot** [comment]
@awaelchli would you verify which one to follow? 

> @awaelchli I think your teammate delivered the wrong message here [#9259 (reply in thread)](https://github.com/Lightning-AI/lightning/discussions/9259#discussioncomment-1858041)

**Comment by awaelchli** [comment]
The answers provided here in this issue are correct. @akihironitta's answer https://github.com/Lightning-AI/lightning/issues/16541#issuecomment-1407662117 should solve the reported problem. In addition, you can see we updated the docs to clarify collectives should be called on all ranks: https://github.com/Lightning-AI/lightning/pull/17046.

**Comment by Aydin-ab** [comment]
I also had a similar issue: I used a custom torchmetrics object to inherit a Metrics class but the code would hang indefinitely when computing the metrics at the end of the epoch (i.e when logging with `self.log("metric_name", metric_object, on_epoch= True`)

The code would hang while running the all_gather operation on the metric's state (a list of lists for each device). At each step, I was appending a new list but some of them did not have the same size and all_gather will hang indefinitely in that case.
To solve it I computed the maximum size possible of my lists and I padded each element to that size before adding them to my metric's state list
